daemon/api.go

```diff
@@ -631,24 +638,58 @@ func (inst *snapInstruction) Agreed(intro, license string) bool {
 	return true
 }
 
-var snappyInstall = snappy.Install
+var snapstateInstall = snapstate.Install
+
+func waitChange(chg *state.Change) error {
+	for {
+		chg.State().Lock()
+		status := chg.Status()
+		err := chg.Err()
+		chg.State().Unlock()
+		switch status {
+		case state.DoneStatus:
+			return nil
+		case state.ErrorStatus:
+			return err
+		}
+		time.Sleep(250 * time.Millisecond)
+	}
+}
 
 func (inst *snapInstruction) install() interface{} {
 	flags := snappy.DoInstallGC
 	if inst.LeaveOld {
 		flags = 0
 	}
-	_, err := snappyInstall(inst.pkg, inst.Channel, flags, inst)
+	state := inst.overlord.StateEngine().State()
+	state.Lock()
+	msg := fmt.Sprintf(i18n.G("Install %q snap"), inst.pkg)
+	if inst.Channel != "stable" {
+		msg = fmt.Sprintf(i18n.G("Install %q snap from %q channel"), inst.pkg, inst.Channel)
+	}
+	chg := state.NewChange("install-snap", msg)
+	ts, err := snapstateInstall(state, inst.pkg, inst.Channel, flags)
+	if err == nil {
+		chg.AddTasks(ts)
+	}
+	state.Unlock()
 	if err != nil {
-		if inst.License != nil && snappy.IsLicenseNotAccepted(err) {
-			return inst.License
-		}
 		return err
 	}
-
-	// TODO: return the log
-	// also to do: commands update their output dynamically, as it changes.
-	return nil
+	inst.overlord.SnapManager().Ensure()
```

Comment:
when we'lll need the contribution of more than one manager this really needs to become a state.EnsureBefore(0) and we should call overlord.Loop() in daemon.Run(?)